clustermesh-apiserver: rework services synchronization to improve performance

giorio94 opened the pull request:
This PR reworks the logic used by the clustermesh-apiserver to synchronize the services and the associated backends to the kvstore. Specifically, it introduces the SyncStore abstraction, which models the operations required to synchronize key/value pairs into a kvstore, and represents a simplified "write-only" version of the SharedStore abstraction currently available. The implementation relies on a workqueue to allow coalescence multiple updates of the same key into one operation only, as well as transparently handling temporary failures through retries. Notably, periodic update of all known keys is not performed, to reduce the load on the kvstore (it was already not performed for CiliumEndpoints and CiliumIdentities, which are also synchronized by the clustermesh-apiserver).

Please, refer to the commit descriptions for additional details about the different changes.

<!-- Description of change -->

```release-note
clustermesh-apiserver: rework services synchronization to improve performance
```

giorio94 committed:
kvstore: add clarification remarks about kvstore backend operations

This commit adds a comment to the delete methods of the interface to
interact with the kvstore backend, to clarify that no error is returned
when the key to be deleted does not exist. Additionally, it corrects
the comment associated with the Update operation, to clarify that it
either creates or updates the key (as the underlying Put operation).

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore: introduce the SyncStore abstraction

This commit introduces the SyncStore abstraction, which models the
operations required to synchronize key/value pairs into a kvstore,
and represents a simplified "write-only" version of the SharedStore
abstraction currently available.

The implementation which is provided in this commit leverages a
workqueue to asynchronously handle the upsert/delete operations
against the kvstore, hence allowing to coalescence multiple updates
of the same key into one operation only, as well as transparently
handling temporary failures through retries.

Additional relevant aspects about the current implementation include:
* Differently from the SharedStore implementation, it does not include a
  periodic update of all known keys, to reduce the load on the kvstore.
* Upsert and delete operations are short-circuited in case they would be
  no-ops according to locally known state (i.e., in case the value did
  not change, or it had already been deleted). This is again to foster
  efficiency, assuming that the kvstore content is not modified externally.
* For consistency with the current implementation, every key is
  associated with a lease. Since it needs to be updated in case of
  restart, we always update the key in case we haven't done that so far,
  even if it might be already present in the kvstore (i.e., we do not
  perform an initial range operation to detect all the keys already
  present that would not require to be updated).
* Stale keys are not explicitly deleted following a restart, but they
  will eventually be removed when the associated lease expires. This is
  consistent with the current SharedStore implementation.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
clustermesh: use SyncStore to synchronize services

This commit updates the service synchronization logic to use the newly
introduced SyncStore implementation, to benefit from the underlying
workqueue to coalescence multiple updates concerning the same service
and automatically handle temporary failures concerning kvstore
operations.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 commented on pkg/kvstore/store/syncstore.go:
```diff
@@ -0,0 +1,190 @@
+// by SyncStore implementations.
+type SyncStoreBackend interface {
+	// Update creates or updates a key.
+	Update(ctx context.Context, key string, value []byte, lease bool) error
```
Another possibility here would be to leverage `UpdateIfDifferent`, which first performs a read operation and then the update if the key does not exist, the value is different, or the lease ID changed. Given that we already know through the local cache whether the entry is up-to-date or not (and we don't perform the update if not necessary), my opinion is that plain `Update` is more efficient, as we save an additional read operation.

marseel replied:
:+1:

marseel commented on pkg/kvstore/store/syncstore.go:
```diff
@@ -0,0 +1,190 @@
+		prefix:    prefix,
+		withLease: true,
+
+		workqueue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
```
I think we will want to make it configurable, right? Also, IIUC default QPS is 10, which might be reasonable for services, but not for other resources in the future.

giorio94 replied:
Yep, I forgot about that. Now it is configurable, and I also changed the `UpsertKey()` and `DeleteKey()` methods to perform a direct insertion in the workqueue (previously they incorrectly went through the rate limiter). Hence, the limiter is only used in case of requeues, since kvstore operations are already rate-limited by a separated rate limiter.

marseel commented on pkg/kvstore/store/syncstore.go:
```diff
@@ -0,0 +1,190 @@
+	w := &wqSyncStore{
+		backend:   backend,
+		prefix:    prefix,
+		withLease: true,
```
We can remove `withLease` parameter here as it will always be true or expose it as parameter?

giorio94 replied:
I've made it possible to configure it through an option, just in case.

marseel commented on pkg/kvstore/store/syncstore.go:
```diff
@@ -0,0 +1,190 @@
+	state     sync.Map /* map[string][]byte --- map[NamedKey.GetKeyName()]Key.Marhal() */
+
+	log *logrus.Entry
+	wg  sync.WaitGroup
```
Do we need this WaitGroup? I couldn't find where `Release` is used except for tests.

giorio94 replied:
I've deleted the `Release()` method and replaced it with a blocking `Run(ctx)` as suggested below.

marseel commented on pkg/kvstore/store/syncstore.go:
```diff
@@ -0,0 +1,190 @@
+// SPDX-License-Identifier: Apache-2.0
```
Probably we want to have some metrics to indicate number of elements in workqueue + how often processing item failed.

giorio94 replied:
I personally agree. I'm wondering though whether to add those in a followup PR, since the clustermesh-apiserver is currently missing the glue logic to expose any metrics.

giorio94 replied:
https://github.com/cilium/cilium/pull/25260/commits/90b759df73db5292a5beb6d8e9efe91991708a5f introduces a new `cilium_kvstore_sync_queue_size` metric to indicate the current size of the queue (per `resource` and `source_cluster`):

```
cilium_kvstore_sync_queue_size{scope="services/v1",source_cluster="left"} 0
```

The number of failures, instead, is already available through the existing `kvstore_operations_duration_seconds` metrics (which is an histogram labeled by `scope`, `action`, `kind` and `outcome`).

Both metrics require https://github.com/cilium/cilium/pull/25316 to be exposed by the clustermesh-apiserver.

marseel commented on pkg/kvstore/store/syncstore.go:
```diff
@@ -0,0 +1,190 @@
+	w.ctx, w.cancel = context.WithCancel(context.Background())
+
+	w.wg.Add(1)
+	go w.syncLoop()
```
I would like to propose a slightly cleaner interface. 
First, we can remove ctx from `wqSyncStore` struct.
We could also add method, something like:
```
func (w *wqSyncStore) Run(ctx context.Context, workers int) {
	defer w.workqueue.ShutDown()

	log.Infof("Starting wgSyncStore")
	defer log.Infof("Shutting down wgSyncStore")

	for i := 0; i < workers; i++ {
		go w.syncLoop(ctx)
	}

	<-ctx.Done()
}
```
WDYT?

giorio94 replied:
Yep, makes definitely sense to me. I've just slightly changed it to remove the worker parameter (configured at creation time) to generalize the interface, and added a waitgroup to wait for all goroutines to terminate before returning.

giorio94 commented on pkg/kvstore/store/syncstore.go:
```diff
@@ -0,0 +1,220 @@
+	UpsertKey(ctx context.Context, key Key) error
+
+	// DeleteLocalKey removes a key from the kvstore.
+	DeleteKey(ctx context.Context, key NamedKey) error
```
Concerning the interface definition, I'm wondering whether to keep the context as parameter for the `UpsertKey` and `DeleteKey` methods. The current workqueue-based implementation ignores it, but it might make sense in a possible synchronous implementations.

giorio94 committed:
kvstore: add queued elements metric to SyncStore

This commit introduces a new `cilium_kvstore_sync_queue_size` metric,
which represents the number of elements queued for synchronization in
the kvstore through the SyncStore abstraction. This metric complements
the ones already existing for the kvstore subsystem, and it is labeled
by `scope` (matching the same format adopted for the other kvstore
metrics) and `source_cluster` (identifying the source cluster of the
information to be synchronized).

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

spacewander commented on pkg/kvstore/store/syncstore.go:
```diff
@@ -0,0 +1,241 @@
+
+	limiter   workqueue.RateLimiter
+	workqueue workqueue.RateLimitingInterface
+	state     sync.Map /* map[string][]byte --- map[NamedKey.GetKeyName()]Key.Marhal() */
```
```suggestion
	state     sync.Map /* map[string][]byte --- map[NamedKey.GetKeyName()]Key.Marshal() */
```

spacewander commented on pkg/kvstore/store/syncstore.go:
```diff
@@ -0,0 +1,241 @@
+	// UpsertKey upserts a key/value pair into the kvstore.
+	UpsertKey(ctx context.Context, key Key) error
+
+	// DeleteLocalKey removes a key from the kvstore.
```
```suggestion
	// DeleteKey removes a key from the kvstore.
```

marseel commented on pkg/kvstore/metrics.go:
```diff
@@ -18,9 +18,9 @@ const (
+func GetScopeFromKey(key string) string {
 	s := strings.SplitN(key, "/", 5)
-	if len(s) != 5 {
+	if len(s) < 4 {
```
I am not following this, why do we change behavior here?

giorio94 replied:
This change is to correctly handle prefixes composed of four items only (e.g., `cilium/state/nodes/v1` -> `nodes/v1`). I've updated the tests to cover that case.

marseel commented on pkg/kvstore/store/syncstore_test.go:
```diff
@@ -0,0 +1,153 @@
+	store = NewWorkqueueSyncStore(backend, "/foo/bar", WSSWithRateLimiter(limiter), WSSWithWorkers(10), WSSWithoutLease()).(*wqSyncStore)
+	require.Equal(t, limiter, store.limiter)
+	require.Equal(t, uint(10), store.workers)
+	require.Equal(t, false, store.withLease)
```
I would prefer to test behavior rather than checking particular fields - this type of tests makes it harder to change names, refactor code etc and any refactor can actually break behavior without test failing - for example, let's say that  `withLease` is not used by SyncStore implementation at all and test will be still passing. 
To test withLease, we could have the assertion that update to backend is done with lease or without, depending on the option. 
For workers, we could have 2 workers and make operation to backend blocking and see if we have two concurrent operations. And so on.

nebril commented on pkg/kvstore/store/syncstore.go:
```diff
@@ -0,0 +1,241 @@
+	// processing this item.
+	defer func() {
+		wss.workqueue.Done(key)
+		wss.queued.Set(float64(wss.workqueue.Len()))
```
Is there a possibility of `processNextItem`, `UpsertKey` and `DeleteKey` running concurrently? If yes, maybe we should use `wss.queued.Inc` and `wss.queued.Dec` instead of `Set` to make sure that gauge value is correct?

giorio94 replied:
`UpsertKey/DeleteKey` might instead run concurrently with `processNextItem`. We cannot use `wss.queued.Inc` and `wss.queued.Dec` though, since the fact that a new element was added to the workqueue does not necessarily mean that the size of the workqueue grew by one (since the same element might be already present). Using `Set` instead, we set the gauge to the current workqueue length (retrieving the length is thread-safe: https://github.com/cilium/cilium/blob/main/vendor/k8s.io/client-go/util/workqueue/queue.go#L144-L148).

nebril reviewed (approved):
LGTM overall, one nitpick left inline about metric value.

marseel reviewed (approved):
Thanks, looks great.

tklauser reviewed (approved):
Neat!

tommyp1ckles reviewed (approved):
Metric changes LGTM

qmonnet reviewed (approved):
Looks good, thanks.

Maybe change the release-note label to `release-note/minor`, given that there's a new metric?
